File: src/main/kotlin/com/huawei/cangjie/psi/CjSuperTypeCallEntry.java
package com.huawei.cangjie.psi;

import com.huawei.cangjie.CjNodeTypes;
import com.huawei.cangjie.psi.stubs.CangJiePlaceHolderStub;
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public class CjSuperTypeCallEntry extends CjSuperTypeListEntry implements CjCallElement {
    public CjSuperTypeCallEntry(@NotNull ASTNode node) {
        super(node);
    }

    public CjSuperTypeCallEntry(@NotNull CangJiePlaceHolderStub<? extends CjSuperTypeListEntry> stub) {
        super(stub, CjStubElementTypes.SUPER_TYPE_CALL_ENTRY);
    }

    @Override
    public <R, D> R accept(@NotNull CjVisitor<R, D> visitor, @Nullable D data) {
        return visitor.visitSuperTypeCallEntry(this, data);
    }

    @NotNull
    @Override
    public CjConstructorCalleeExpression getCalleeExpression() {
        return getRequiredStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE);
    }

    @Override
    public @Nullable CjValueArgumentList getValueArgumentList() {
        return (CjValueArgumentList) findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST);

    }


    @Override
    public CjTypeReference getTypeReference() {
        return getCalleeExpression().getTypeReference();
    }



}

File: src/main/kotlin/com/huawei/cangjie/psi/CjSuperTypeCallEntry.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.CjNodeTypes
import com.huawei.cangjie.psi.stubs.CangJiePlaceHolderStub
import com.huawei.cangjie.psi.stubs.elements.CjStubElementTypes
import com.intellij.lang.ASTNode

class CjSuperTypeCallEntry : CjSuperTypeListEntry, CjCallElement {
    constructor(node: ASTNode) : super(node)

    constructor(stub: CangJiePlaceHolderStub<out CjSuperTypeListEntry >) : super(
        stub,
        CjStubElementTypes.SUPER_TYPE_CALL_ENTRY
    )

    override fun <R, D> accept(visitor: CjVisitor<R, D>, data: D?): R {
        return visitor.visitSuperTypeCallEntry(this, data)
    }

    override val calleeExpression : CjConstructorCalleeExpression get()   {
        return getRequiredStubOrPsiChild(CjStubElementTypes.CONSTRUCTOR_CALLEE)
    }

    override val lambdaArguments : List<CjLambdaArgument> get(){
        return emptyList()
    }

    override val typeArguments : List<CjTypeProjection> get(){
        val typeArgumentList = typeArgumentList ?: return emptyList()
        return typeArgumentList.arguments
    }

    override val typeArgumentList : CjTypeArgumentList?   = null

    override val valueArgumentList : CjValueArgumentList? get(){
        return findChildByType(CjNodeTypes.VALUE_ARGUMENT_LIST)
    }

    override val valueArguments : List<ValueArgument > get(){
        val list = valueArgumentList
        return list?.arguments ?: emptyList<CjValueArgument>()
    }


    override val typeReference : CjTypeReference? get(){
        return calleeExpression.typeReference
    }
}
